### PR TITLE
fix(StatusPopupMenu): Refactoring

### PR DESCRIPTION
**Pull request by igor-sirotin**
Part of https://github.com/status-im/status-desktop/issues/7232
As continuation of https://github.com/status-im/StatusQ/pull/894.

⚠️  **NOTE:**
This is a PR to a temporary branch `fix/status-menu`
After this PR is merged, I will update and test the main application. And create a PR to `master`.

1. All Menu components are refactored.
2. Some UI improvements made:
    - Highlight menu when submenu opened
    - Proper colors for disabled menu items
3. I'm also planning to rename the components as following in a separate PR after testing with status-desktop

| Old name | New name |
| ---------- | ----------- |
| StatusPopupMenu | StatusMenu |
| StatusMenuItem | StatusAction |
| StatusMenuItemDelegate | StatusMenuItem |

### StatusQ checklist

- [ ] add documentation if necessary (new component, new feature)
- [x] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- [x] test changes in both light and dark theme?

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

https://user-images.githubusercontent.com/25482501/203008348-e0a7e043-8168-41b3-97a6-2efea8d02269.mov

### CODE REVIEW

File: ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml [hunk 1]
```diff
@@ -7,187 +7,164 @@ import StatusQ.Components 0.1
...
         }
     }
 
     MouseArea {
-        id: sensor
         anchors.fill: parent
         cursorShape: Qt.PointingHandCursor
-        hoverEnabled: statusPopupMenuItem.action.enabled
+        hoverEnabled: root.enabled
         onPressed: mouse.accepted = false
```
Comment (caybro) [inline ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml:168]: I'd remove this onPressed handler; you can achieve the same with setting `acceptedButtons: Qt.NoButton`

File: ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml [hunk 2]
```diff
@@ -7,187 +7,164 @@ import StatusQ.Components 0.1
...
+        text: root.text
+        color: !root.enabled ? Theme.palette.baseColor1
+                             : d.isDangerIcon ? Theme.palette.dangerColor1 : Theme.palette.directColor1
+
+        font.pixelSize: d.fontSettings.pixelSize
+        font.bold: d.fontSettings.bold
+        font.italic: d.fontSettings.italic
         elide: Text.ElideRight
-        visible: statusPopupMenuItem.action.enabled
+        visible: true
```
Comment (caybro) [inline ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml:140]: This is the default

**Review by caybro — COMMENTED** [review]
Code looks good, just some minor nits

File: ui/StatusQ/src/StatusQ/Popups/StatusPopupMenu.qml [hunk 3]
```diff
@@ -10,21 +10,32 @@ import StatusQ.Popups 0.1
...
     id: root
+
     closePolicy: Popup.CloseOnPressOutside | Popup.CloseOnEscape
     topPadding: 8
     bottomPadding: 8
     bottomMargin: 16
 
-    property int menuItemCount: 0
-    property var subMenuItemIcons: []
+    property StatusAssetSettings assetSettings: StatusAssetSettings {
```
Comment (osmaczko) [inline ui/StatusQ/src/StatusQ/Popups/StatusPopupMenu.qml:19]: IMO it doesn't belong here. It should be defined by the delegate.

File: ui/StatusQ/src/StatusQ/Popups/StatusPopupMenu.qml [hunk 4]
```diff
@@ -10,21 +10,32 @@ import StatusQ.Popups 0.1
...
+        width: 18
+        height: 18
+        rotation: 0
+        isLetterIdenticon: false
+        isImage: false
+        color: "transparent"
+    }
+
+    property StatusFontSettings fontSettings: StatusFontSettings {}
```
Comment (osmaczko) [inline ui/StatusQ/src/StatusQ/Popups/StatusPopupMenu.qml:28]: Same as above.

File: ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml [hunk 5]
```diff
@@ -7,187 +7,164 @@ import StatusQ.Components 0.1
...
-        if (!!subMenu) {
-            subMenuIndex = statusPopupMenu.menuItemCount
-            statusPopupMenu.menuItemCount += 1
+    implicitHeight: menu.hideDisabledItems && !enabled ? 0 : 38
+    objectName: action ? action.objectName : "StatusMenuItemDelegate"
+
+    spacing: 4
+    horizontalPadding: 8
+
+    readonly property string logObjectName: "StatusMenuItemDelegate [%1, %2]".arg(this).arg(text)
```
Comment (osmaczko) [inline ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml:18]: Leftover?

File: ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml [hunk 6]
```diff
@@ -7,187 +7,164 @@ import StatusQ.Components 0.1
...
+
+    QtObject {
+        id: d
+
+        readonly property bool isSubMenu: !!root.subMenu
+        readonly property bool isStatusSubMenu: isSubMenu && (root.subMenu instanceof StatusPopupMenu)
+        readonly property bool subMenuOpened: isSubMenu && root.subMenu.opened
+        readonly property bool hasAction: !!root.action
+        readonly property bool isStatusAction: d.hasAction && (root.action instanceof StatusMenuItem)
+        readonly property bool isDangerIcon: d.isStatusAction && root.action.type === StatusMenuItem.Type.Danger
```
Comment (osmaczko) [inline ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml:28]:
```suggestion
        readonly property bool isStatusDangerAction: d.isStatusAction && root.action.type === StatusMenuItem.Type.Danger
```

File: ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml [hunk 7]
```diff
@@ -7,187 +7,164 @@ import StatusQ.Components 0.1
...
 
     background: Rectangle {
         color: {
-            if (statusPopupMenuItem.hovered) {
-                return statusPopupMenuItem.action.type === StatusMenuItem.Type.Danger ? Theme.palette.dangerColor3 : Theme.palette.statusPopupMenu.hoverBackgroundColor
-            }
-            return "transparent"
+            if (!root.hovered && !d.subMenuOpened)
+                return "transparent"
+            if (d.isStatusAction && root.action.type === StatusMenuItem.Type.Danger)
```
Comment (osmaczko) [inline ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml:158]:
```suggestion
            if (d.isDangerIcon)
```

File: ui/StatusQ/sandbox/pages/StatusPopupMenuPage.qml [hunk 8]
```diff
@@ -51,7 +51,7 @@ GridLayout {
 
     StatusPopupMenu {
         id: complexMenu
-        subMenuItemIcons: [{ icon: 'info' }]
+//        subMenuItemIcons: [{ icon: 'info' }]
```
Comment (osmaczko) [inline ui/StatusQ/sandbox/pages/StatusPopupMenuPage.qml:54]: :police_officer:

**Review by osmaczko — COMMENTED** [review]
Nice delegate simplification :+1:  left some comments

File: ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml (hunk 6 above)
Comment (caybro) [reply ui/StatusQ/src/StatusQ/Popups/StatusMenuItemDelegate.qml:28]: Oh yeah, QtC offers a great variety of QML refactoring tools :smiling_imp: not...

**Review by osmaczko — APPROVED** [review]

**Review by caybro — APPROVED** [review]

**Pull request merged**